### DISCUSSION TITLE
Limit size of feature per sample

### DISCUSSION THREAD

**Discussion by qbhan** [mitsuba-renderer/mitsuba3#535] [Q&A] [answered]
## Summary
I am trying to create a dataset for various Monte Carlo Denoisers.
For instance, to train a Sample-based Monte Carlo Denoising (Gharbi et al. 2019) with Path-Manifold Module (Cho et al.2021) I need about 51+36 = 87 channels of features per sample.

I successfully managed to get 51 channels for Sample-based Monte Carlo Denoising. However, I'm facing an error as I try to get more than 75 channels. Below are the printed 76 channels I want to get from each sample and the error message.
```shell
(mitsuba) kyubeom@793f1e47e7e1:~/mitsuba3_playground$ python integrators.py 
['normal:3f', 'depth:1f', 'albedo:3f', 'normal_diff:3f', 'depth_diff:1f', 'albedo_diff:3f', 'visibility_b0:1f', 'visibility_b1:1f', 'direction_b0:2f', 'direction_b1:2f', 'direction_b2:2f', 'direction_b3:2f', 'direction_b4:2f', 'direction_b5:2f', 'probability_b0:4f', 'probability_b1:4f', 'probability_b2:4f', 'probability_b3:4f', 'probability_b4:4f', 'probability_b5:4f', 'material_b0:1i', 'material_b1:1i', 'material_b2:1i', 'material_b3:1i', 'material_b4:1i', 'material_b5:1i', 'throughput_b0:3f', 'throughput_b1:3f', 'throughput_b2:3f', 'throughput_b3:3f', 'throughput_b4:3f', 'throughput_b5:3f']


Critical Dr.Jit compiler failure: jit_var(r3215): unknown variable!
Aborted (core dumped)
```

I'm wondering if there is any memory limit that a loop of a single sample can handle.
I would really appreciate some help on how to solve the problem.

## System configuration

System information:

  OS: Ubuntu 20.04
  CPU: Intel(R) Core(TM) i9-10980XE CPU @ 3.00GHz
  GPU: RTX3090
  Python version: 3.11.0
  LLVM version: None
  CUDA version: 11.6
  NVidia driver: 510.73.05

  Dr.Jit version: 0.4.0
  Mitsuba version: 3.2.0
     Compiled with: pip install mitsuba
     Variants compiled: 
scalar_rgb
scalar_spectral
cuda_ad_rgb
llvm_ad_rgb

## Description

I added a list of features as "aovs" on a simple path tracer with NEE and MIS. The name of features are given in the initialization stage. 
The code below shows how the loop of sample() is configured.

```
def sample(self, scene: mi.Scene,
               sampler: mi.Sampler, 
               ray_: mi.RayDifferential3f, 
               medium: mi.Medium = None, 
               active: bool = True
               ):
        bsdf_ctx = mi.BSDFContext()


        # --------------------- Configure loop state ----------------------

        ray = mi.Ray3f(ray_)
        prev_si = dr.zeros(mi.SurfaceInteraction3f)

        depth = mi.UInt32(0)                            # Depth of current vertex
        L = mi.Spectrum(0.0)                            # Radiance accumulator
        β = mi.Spectrum(1.0)                            # Path throughput weight
        η = mi.Float(1)                                 # Index of refraction
        mis_em = mi.Float(1)                            # Emitter MIS weight
        active = mi.Bool(active)                        # Active SIMD lanes
        
        aovs = []
        for i in range(len(self.aov_list)):
            if '1f' in self.aov_list[i]: aovs.append(mi.Vector1f(0.0))
            elif '2f' in self.aov_list[i]: aovs.append(mi.Vector2f(0.0))
            elif '3f' in self.aov_list[i]: aovs.append(mi.Vector3f(0.0))
            elif '4f' in self.aov_list[i]: aovs.append(mi.Vector4f(0.0))
            elif '1i' in self.aov_list[i]: aovs.append(mi.Vector1i(0))


        direction = mi.Vector2f(0.0)

        first_non_specular = mi.Bool(False)
        depth_diff = mi.Float(0.0)
        β_diff = mi.Spectrum(1.0)
        L_diff = mi.Spectrum(0.0)

        loop = mi.Loop(name="Path Tracing", state=lambda: (
            sampler, ray, prev_si, depth, L, β, η, mis_em, active, aovs, first_non_specular, depth_diff, β_diff, L_diff))

        loop.set_max_iterations(self.max_depth)
```

Even though I assign dummy values like mi.Vector3f(1.0) still the error occurs. 
I really appreciate the help!

**Comment by njroussel** [comment, marked as answer]
Hi @qbhan 

Could you try compiling the project yourself rather than using the `pip install` version? I don't think there is any limit to the AOVs, but rather that you're encountering the same issue as here: https://github.com/mitsuba-renderer/mitsuba3/issues/482. Compiling Mitsuba on your own using `master` might fix your problem.

**Reply by qbhan** [reply]
I just complied it manually and the error disappeared. Thank you!
